File: WoofAndroid/data/datasource/src/testFixtures/kotlin/com/perrystreet/woof/datasource/dogs/DogsDataSourceFactory.kt
package com.perrystreet.woof.datasource.dogs

import com.perrystreet.woof.dto.dog.DogDTO
import org.koin.test.KoinTest
import org.koin.test.inject

class DogsDataSourceFactory : KoinTest {
    private val dataSource: FakeDogsDataSource by inject()

    fun withDogs(count: Int) = apply {
        dataSource.dogs = (1..count).map { index ->
            DogDTO(id = index.toLong(), name = "Dog $index", photoUrl = "file:///android_asset/dogs/golden_retriever_01.jpg")
        }
    }

    fun withDogsError() = apply {
        dataSource.getDogsError = IllegalStateException("Could not load dogs")
    }

    fun withoutErrors() = apply {
        dataSource.getDogsError = null
        dataSource.getDogProfileError = null
    }

    fun withDogProfileError() = apply {
        dataSource.getDogProfileError = IllegalStateException("Could not load profile")
    }
}

File: WoofIOS/SwiftPackages/Data/DataSource/Sources/DataSourceFakes/Dogs/DogsDataSourceFactory.swift
import DTO
import DataSource
import Swinject
import SwinjectAutoregistration

public final class DogsDataSourceFactory {
    private let dataSource: FakeDogsDataSource

    public init(_ container: Container) {
        dataSource = container~>
    }

    @discardableResult
    public func withDogs(count: Int) -> Self {
        dataSource.dogs = (1...count).map { index in
            DogDTO(id: index, name: "Dog \(index)", photoUrl: "woof://dogs/golden_retriever_01.jpg")
        }
        return self
    }

    @discardableResult
    public func withDogsError() -> Self {
        dataSource.getDogsError = .unavailable("Could not load dogs")
        return self
    }

    @discardableResult
    public func withoutErrors() -> Self {
        dataSource.getDogsError = nil
        dataSource.getDogProfileError = nil
        return self
    }

    @discardableResult
    public func withDogProfileError() -> Self {
        dataSource.getDogProfileError = .unavailable("Could not load profile")
        return self
    }
}
